### DISCUSSION TITLE
math_opt

### DISCUSSION THREAD

**Discussion by C8504** [google/or-tools#2935] [Linear Solver questions] [answered]
Hi Experts,

Could I know what is the benifits in math_opt compare with linear_solver?
Is it more compact in  terms of memory or faster create and query models?

Thanks
John

**Comment by lperron** [comment, marked as answer]
1) Math Opt is experimental
2) it is meant to be used for advanced MIP solving (lagrangian relaxation , column generation...)
3) it is not meant to be faster/smaller for single solve

**Reply by C8504** [reply]
Thanks Lperron.

**Comment by rma350** [comment]
A few things to add:

- as lperron said, the focus is on supporting more features beyond a simple solve, e.g. incrementalism, callbacks, rays, basis, duality, primal/dual infeasible solutions, return stats, parameters, etc.
- math_opt separates the model from the solver, and allows you to define the model without any C++ dependency for non C++ users.
- There are different engineering choices that will hopefully make it easier to add new solvers and maintain the code.
- There will be performance differences, hopefully more often better than worse. A wider range of queries are supported in O(1). However, performance of the modeling layer is usually not the bottleneck unless: (a) you are solving with a callback, (b) you are incrementally solving LPs, or (c) your LP is easy, relative to its size.
- The version on github is behind head and parts of the API have changed. You are welcome to experiment with it, but you will face breaking changes.

**Reply by C8504** [reply]
Thanks Ross. 
Actually, the benders sample in math_opt is why I am curious about this.
Hopefully this part of ortools works well and makes big impact compare with math solver interfaces.
Recently, https://www.gravityopt.com/ in coinor did a good job on C++ modeling domain, don't know if you know this.

John
